url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/methodology/History.java

```diff
@@ -0,0 +1,380 @@
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new NumberOfRounds(new IntegerRelation(), new NoScale<Integer>()),
+				new SizeOfMessages(new SizeOfMessagesRelation(), new NoScale<Integer>()),
+				new CryptoCalls(new IntegerRelation(), new NoScale<Integer>()),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+				new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()),
+				new YearOfPublication(new IntegerRelation(), new NoScale<Integer>()),
+		};
+		
+		ParetoFrontier[] frontiers = ParetoFrontier.computeAllParetoFrontiers(protocols, attributes);
+		
+		saveInDiskTheFrontiers(frontiers, "history.obj");
+		
+		
+		//printLatexTable(protocols, attributes, frontiers, 0, 8, "history_table.tex");
+		//printHistory(protocols, attributes, frontiers);
+
+		//TreeBasedProtocol tree = new TreeBasedProtocol();
+		//System.out.println(tree.getMemory(63));
+		//KimAndAvoineProtocol kim = new KimAndAvoineProtocol(0.25, 128);
+		//DominatingRelation relation = new DominatingRelation(order, protocols, attributes, 64);
+		//DBProtocol[] protocols = new DBProtocol[]{tree, kim};
+		//relation.printInfoOfDomination(tree, kim, 63);
+	}
+
+	public static void saveInDiskTheFrontiers(ParetoFrontier[] frontiers,
+			String file) throws FileNotFoundException, IOException {
+		ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(file));
+		out.writeObject(frontiers);
+		out.close();
+	}
+
+	/*Trujillo- Apr 4, 2014
+	 * This method takes MAX_N and builds the history from n = 1 to n = MAX_N. This history is basically how the 
+	 * pareto frontier evolves with n.*/
+	public static void printHistory(DBProtocol[][] protocols,  
+			Attribute[] attributes, ParetoFrontier[] frontiers){
+		long total = 0; 
+		for (int i = 0; i < frontiers.length; i++) {
+			total += frontiers[i].getProtocols().length;
+		}
+		System.out.println("There are "+total+" nondominated protocol instances");
+		ParetoFrontier before = null;
+		for (int i = 0; i < protocols.length; i++) {
+			//System.out.println("Computing pareto frontier for e = "+i);
+			System.out.println("");
+			ParetoFrontier frontier = frontiers[i];
+			System.out.println("");
+			
+			if (before != null){
+				basicInfo(frontier.getFrontier(), before.getFrontier());
+				System.out.println("");
+				provideReasons(frontier, before, attributes);			
+				System.out.println("");
+			}else 
+				basicInfo(frontier.getFrontier()); 
+			before = frontier;
+		}
+	}
+	
+	/*Trujillo- Apr 5, 2014
+	 * Print the history in a latex table format*/
+	public static void printLatexTable(ParetoFrontier[] frontiers, 
+			int remains, int module, String name) throws IOException{
+		if (frontiers.length == 0) return;
+		FileWriter writer = new FileWriter(name);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *writer*. The resource is referenced at lines 116, 142, 144. The resource is closed at line 145. There are other execution paths that do not contain closure statements, for example, when Latex.appendTableFooter throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).